fix(service-worker): make `ngsw.json` generation deterministic and correct

gkalpak opened the pull request:
Previously, all asset-groups from `ngsw-config.json` were processed in parallel. For each asset-group, we retrieved all files for the current build, filtered out files that were already matched by other asset-groups, determined which of the remaining files belonged to the current asset-group and generated entries for the `ngsw.json` manifest. This process was susceptible to race conditions when there were files that would be matched by multiple asset-groups. This made the generation of the `ngsw.json` manifest non-deterministic and violated the rule that each file would belong to the first asset-group that matched it (based on the asset-groups' order of appearance in `ngsw-config.json`), thus leading to broken ServiceWorker behavior.

This commit fixes it by ensuring that the generation process is deterministic and that asset-groups are processed in the proper order.

NOTE 1:
The generation process has been broken since the beginning, but we have only noticed this recently. This is possibly related to the CLI's switching from a virtual file system host (which has more consistent timing characteristics) to the Node.js built-in `fs.promises` in angular/angular-cli@d3bc530c103849ca83742ab63eea7709d9af613d.

NOTE 2:
This commit also ensures that files in the `ngsw.json` hash-table are in alphabetic order. Previously, the files were added to the hash-table in blocks corresponding to each asset-group.
This change is not necessary (i.e. the order of keys in the hash-table makes no difference in behavior), but it makes it easier to scan for a file (for example, for debugging purposes).

##
_**NOTE:**_
_Once AIO is updated to a version that includes this fix, remove [this work-around][1]._

[1]: https://github.com/angular/angular/pull/43686/commits/3d7b8410f2b259698833cde16fe897ebf8f58136

mary-poppins commented:
You can preview 9b34c44 at https://pr43679-9b34c44.ngbuilds.io/.

mary-poppins commented:
You can preview a98238e at https://pr43679-a98238e.ngbuilds.io/.

petebacondarwin commented on packages/service-worker/config/src/generator.ts:
```diff
@@ -45,36 +45,44 @@ export class Generator {
-        hashTable[joinUrls(this.baseHref, file)] = hash;
-      }, Promise.resolve());
+    // Compute hashes for all matched files and add them to the hash-table.
+    const allMatchedFiles = ([] as string[]).concat(...Array.from(filesPerGroup.values())).sort();
```
I think the concat is unnecessary here since the `values()` method will return a new array each time:

```suggestion
    const allMatchedFiles = Array.from(filesPerGroup.values()).sort();
```

gkalpak replied:
`Array.from(filesPerGroup.values())` is an array of arrays. The `.concat()` is for flattening it :grin:

petebacondarwin replied:
Ahah! I didn't notice the `...`. Fair enough

petebacondarwin commented on packages/service-worker/config/src/generator.ts:
```diff
@@ -45,36 +45,44 @@ export class Generator {
-      }, Promise.resolve());
+    // Compute hashes for all matched files and add them to the hash-table.
+    const allMatchedFiles = ([] as string[]).concat(...Array.from(filesPerGroup.values())).sort();
+    const allMatchedHashes = await Promise.all(allMatchedFiles.map(file => this.fs.hash(file)));
```
Not related to this PR...

The `hash()` implementations that I have seen in the FW code are all synch under the hood.
Perhaps there are other implementations out there that are async?

gkalpak replied:
Sorry, I am not following 😅
I am sure there async implementation, but what difference does it make? 🤔

petebacondarwin replied:
It makes not much difference except that making that interface `async` causes the async-ness to bubble up through all usage, which is unnecessary.

petebacondarwin reviewed (changes requested):
I feel like we need some kind of test to prove out this fix??

gkalpak commented:
> I feel like we need some kind of test to prove out this fix??

Sigh 😞 There used to be a time when my word was enough 😛
Added a test 😃

mary-poppins commented:
You can preview 588a00e at https://pr43679-588a00e.ngbuilds.io/.

petebacondarwin commented on packages/service-worker/config/test/generator_spec.ts:
```diff
@@ -128,6 +128,102 @@ describe('Generator', () => {
+  it('assigns files to the first matching asset-group (unaffected by file-system access delays)',
+     async () => {
+       const fs = new MockFilesystem({
+         '/index.html': 'This is a test',
```
You don't say 😛 ?

petebacondarwin commented on packages/service-worker/config/test/generator_spec.ts:
```diff
@@ -128,6 +128,102 @@ describe('Generator', () => {
+             resources: {files: ['/bar/**/*.js']},
+           },
+           {
+             name: 'group-fallback',
+             resources: {files: ['/**/*.js']},
+           },
```
So this is the key group in this test, right? Since it matches files that should have already been matched by the previous groups.

gkalpak replied:
Yes.

petebacondarwin reviewed (approved):
Thanks for the test ❤️ ❤️ ❤️ ❤️

gkalpak committed:
fix(service-worker): make `ngsw.json` generation deterministic and correct

Previously, all asset-groups from `ngsw-config.json` were processed in
parallel. For each asset-group, we retrieved all files for the current
build, filtered out files that were already matched by other
asset-groups, determined which of the remaining files belonged to the
current asset-group and generated entries for the `ngsw.json` manifest.
This process was susceptible to race conditions when there were files
that would be matched by multiple asset-groups. This made the generation
of the `ngsw.json` manifest non-deterministic and violated the rule that
each file would belong to the first asset-group that matched it (based
on the asset-groups' order of appearance in `ngsw-config.json`), thus
leading to broken ServiceWorker behavior.

This commit fixes it by ensuring that the generation process is
deterministic and that asset-groups are processed in the proper order.

NOTE 1:
The generation process has been broken since the beginning, but we have
only noticed this recently. This is possibly related to the CLI's
switching from a virtual file system host (which has more consistent
timing characteristics) to the Node.js built-in `fs.promises` in
angular/angular-cli@d3bc530c103849ca83742ab63eea7709d9af613d.

NOTE 2:
This commit also ensures that files in the `ngsw.json` hash-table are in
alphabetic order. Previously, the files were added to the hash-table in
blocks corresponding to each asset-group.
This change is not necessary (i.e. the order of keys in the hash-table
makes no difference in behavior), but it makes it easier to scan for a
file (for example, for debugging purposes).

gkalpak committed:
fixup! fix(service-worker): make `ngsw.json` generation deterministic and correct

mary-poppins commented:
You can preview be62c33 at https://pr43679-be62c33.ngbuilds.io/.

dylhunn commented:
This PR was merged into the repository by commit fddb50b597ceaacc77f8412dda29f0ae2bd3efbe.
